File: api/routes/themes.py
"""Theme API routes."""

from fastapi import APIRouter, HTTPException
from typing import List

from src.maptoposter.themes.loader import (
    get_available_themes,
    load_theme,
    get_all_themes_with_details
)
from src.maptoposter.exceptions import ThemeNotFoundError
from api.schemas.poster import ThemeInfo, ThemeListResponse

router = APIRouter()


@router.get("", response_model=ThemeListResponse)
async def list_themes():
    """Get list of all available themes."""
    themes_data = get_all_themes_with_details()

    themes = []
    for t in themes_data:
        themes.append(ThemeInfo(
            id=t.get('id', 'unknown'),
            name=t.get('name', t.get('id', 'Unknown')),
            description=t.get('description'),
            bg=t.get('bg', '#FFFFFF'),
            text=t.get('text', '#000000'),
            road_motorway=t.get('road_motorway', '#000000'),
            road_primary=t.get('road_primary', '#333333'),
            water=t.get('water', '#C0C0C0'),
            parks=t.get('parks', '#F0F0F0')
        ))

    return ThemeListResponse(themes=themes)


@router.get("/{theme_name}", response_model=ThemeInfo)
async def get_theme(theme_name: str):
    """Get details of a specific theme."""
    try:
        theme = load_theme(theme_name)
        return ThemeInfo(
            id=theme_name,
            name=theme.get('name', theme_name),
            description=theme.get('description'),
            bg=theme.get('bg', '#FFFFFF'),
            text=theme.get('text', '#000000'),
            road_motorway=theme.get('road_motorway', '#000000'),
            road_primary=theme.get('road_primary', '#333333'),
            water=theme.get('water', '#C0C0C0'),
            parks=theme.get('parks', '#F0F0F0')
        )
    except ThemeNotFoundError:
        raise HTTPException(status_code=404, detail=f"Theme '{theme_name}' not found")

File: maptoposter-rs/src/api/handlers/themes.rs
use std::sync::Arc;

use axum::{
    extract::{Path, State},
    Json,
};

use crate::api::models::{ThemeInfo, ThemeListResponse};
use crate::api::state::AppState;
use crate::error::{AppError, Result};
use crate::themes::loader::{load_theme, load_themes};

/// List all available themes
pub async fn list_themes(State(state): State<Arc<AppState>>) -> Json<ThemeListResponse> {
    let themes_map = load_themes(&state.config.themes_dir);

    // Collect and sort themes by name for consistent ordering
    let mut themes: Vec<ThemeInfo> = themes_map
        .into_iter()
        .map(|(name, theme)| ThemeInfo {
            id: name.clone(),
            name,
            description: theme.get("description").and_then(|v| v.as_str()).map(String::from),
            bg: theme.get("bg").and_then(|v| v.as_str()).unwrap_or("#FFFFFF").to_string(),
            text: theme.get("text").and_then(|v| v.as_str()).unwrap_or("#000000").to_string(),
            gradient_color: theme.get("gradient_color").and_then(|v| v.as_str()).unwrap_or("#FFFFFF").to_string(),
            water: theme.get("water").and_then(|v| v.as_str()).unwrap_or("#C0C0C0").to_string(),
            parks: theme.get("parks").and_then(|v| v.as_str()).unwrap_or("#F0F0F0").to_string(),
            road_motorway: theme.get("road_motorway").and_then(|v| v.as_str()).map(String::from),
            road_primary: theme.get("road_primary").and_then(|v| v.as_str()).map(String::from),
            road_default: theme.get("road_default").and_then(|v| v.as_str()).map(String::from),
        })
        .collect();

    // Sort alphabetically by id for consistent UI ordering
    themes.sort_by(|a, b| a.id.cmp(&b.id));

    let count = themes.len();
    Json(ThemeListResponse { themes, count })
}

/// Get a specific theme by name
pub async fn get_theme(
    State(state): State<Arc<AppState>>,
    Path(name): Path<String>,
) -> Result<Json<ThemeInfo>> {
    let theme = load_theme(&state.config.themes_dir, &name)
        .ok_or_else(|| AppError::ThemeNotFound(name.clone()))?;

    Ok(Json(ThemeInfo {
        id: name.clone(),
        name,
        description: theme.get("description").and_then(|v| v.as_str()).map(String::from),
        bg: theme.get("bg").and_then(|v| v.as_str()).unwrap_or("#FFFFFF").to_string(),
        text: theme.get("text").and_then(|v| v.as_str()).unwrap_or("#000000").to_string(),
        gradient_color: theme.get("gradient_color").and_then(|v| v.as_str()).unwrap_or("#FFFFFF").to_string(),
        water: theme.get("water").and_then(|v| v.as_str()).unwrap_or("#C0C0C0").to_string(),
        parks: theme.get("parks").and_then(|v| v.as_str()).unwrap_or("#F0F0F0").to_string(),
        road_motorway: theme.get("road_motorway").and_then(|v| v.as_str()).map(String::from),
        road_primary: theme.get("road_primary").and_then(|v| v.as_str()).map(String::from),
        road_default: theme.get("road_default").and_then(|v| v.as_str()).map(String::from),
    }))
}
